### ISSUE TITLE
Overriding compile to support other template formats

### ISSUE THREAD

**Issue by AndrewLeedham** [wycats/handlebars.js#1547]
As per #1534:

The following JS Fiddle shows how we are overriding the handlebars compile function in order to allow mustache templates to be optionally used: https://jsfiddle.net/AndrewLeedham/1Ljzkneh/28/

So essentially it is overriding compile and checking if a partial has its handler set to `mustache` exists given the name of the current template being rendered. If so it uses `Mustache.render`, otherwise it carries on with the normal Handlebars compile. All the partials have to be registered to handlebars, otherwise the callback returned from Handlebars.compile is never run.

**Comment by nknapp** [comment]
I have to think about this. This code looks hacky to me, but I am not sure about the alternatives.

You *could* register a "compiled" partial like in https://jsfiddle.net/hq413g0a/, but this is equally undocumented and there is no test in the spec that ensures this behaviour. It just works, because the partial is replaced by its compiled function in the partials-array

It would be cleaner to register a helper for the mustache partials, but I can see that this would require a lot of adaption in your code.

**Comment by AndrewLeedham** [comment]
I agree this is hacky. We plan to move away from the pattern eventually, but as you say it is a solution that works now. Just thought I'd share it so you are aware of how handlebars may be used in some cases, so you can use that information for future changes.

Not sure what you referring to with a "compiled" partial, is that essentially registering a partial as a function?

Yes a helper wouldn't be ideal in our use case.

**Comment by nknapp** [comment]
You usually register a string as partial, but when you include the partial for the first time, it is replaced by the compiled function. What I have done is registering a function directly. Something like that also happens when you precompile partials using the `handlebars` binary.

https://jsfiddle.net/xrdonyc5/2/

**Comment by nknapp** [comment]
@wycats I would like to know if you would consider registering a function as a partial is considered part of the API. It **is** possible at the moment, but undocumented.
